validator/iam_allow_ban_roles.rego

```diff
@@ -0,0 +1,59 @@
+#
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPIAMAllowBanRolesConstraintV1
+
+import data.validator.gcp.lib as lib
+
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	lib.get_constraint_params(constraint, params)
+	asset := input.asset
+
+	binding := asset.iam_policy.bindings[_]
+	role := binding.role
+
+	matches_found = {r | r := role; glob.match(params.roles[_], [], r)}
+
+	mode := lib.get_default(params, "mode", "allow")
+
+	target_match_count(mode, desired_count)
+	count(matches_found) != desired_count
```

Comment:
If the user has two matches in parameters.roles (eg,["roles/viewer", "roles/*"]) this will fail on legal inputs for the allow case.  I think you need something like (match == "allow") == (0 < count(matches_found)).

Reply:
This situation would not happen, because these two lines
```
binding := asset.iam_policy.bindings[_]
role := binding.role
```
determines that we are checking the roles one by one. See [here](https://www.openpolicyagent.org/docs/v0.10.7/how-does-opa-work/) for the following quote

> Expressions can reference elements in a collection using the [_] and [<variable>] syntax. OPA knows to evaluate such queries by iterating over each element in the corresponding collection.

So in `matches_found = {r | r := role; glob.match(params.roles[_], [], r)}`, there is always only one value for `r` in each iteration. Therefore, value of `count(matches_found)` can only be 1 or 0.

Reply:
oh, my mistake, I misread the {r | r ...} as [r | r ...]